nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/util/FlowDifferenceFilters.java

```diff
@@ -547,4 +577,302 @@ private static boolean isParameterContextChange(final FlowDifference flowDiffere
+                }
+            }
+        }
+
+        Set<String> serviceIdsWithMatchingAdditions = Collections.emptySet();
+        if (!serviceIdsReferencedByNewProperties.isEmpty()) {
+            serviceIdsWithMatchingAdditions = differences.stream()
+                    .filter(diff -> diff.getDifferenceType() == DifferenceType.COMPONENT_ADDED)
+                    .map(FlowDifferenceFilters::extractControllerServiceIdentifier)
+                    .filter(Objects::nonNull)
+                    .filter(serviceIdsReferencedByNewProperties::contains)
+                    .collect(Collectors.toCollection(HashSet::new));
+        }
+
+        final Set<FlowDifference> parameterizedPropertyRenameDifferences = new HashSet<>();
+        for (final Map.Entry<String, List<PropertyDiffInfo>> entry : parameterizationRemovalsByComponent.entrySet()) {
+            final String componentId = entry.getKey();
+            final List<PropertyDiffInfo> removals = entry.getValue();
+            final List<PropertyDiffInfo> additions = new ArrayList<>(parameterizedAddsByComponent.getOrDefault(componentId, Collections.emptyList()));
+            if (additions.isEmpty()) {
+                continue;
+            }
+
+            for (final PropertyDiffInfo removalInfo : removals) {
+                final Optional<String> removalValue = removalInfo.propertyValue();
+                if (removalValue.isPresent() && !isParameterReference(removalValue.get())) {
+                    continue;
+                }
+
+                PropertyDiffInfo matchingAddition = null;
+                for (final Iterator<PropertyDiffInfo> iterator = additions.iterator(); iterator.hasNext();) {
+                    final PropertyDiffInfo additionInfo = iterator.next();
+                    final Optional<String> additionValue = additionInfo.propertyValue();
+                    if (additionValue.isPresent() && !isParameterReference(additionValue.get())) {
+                        continue;
+                    }
+
+                    if (valuesMatch(removalValue, additionValue)) {
+                        matchingAddition = additionInfo;
+                        iterator.remove();
+                        break;
+                    }
+                }
+
+                if (matchingAddition != null) {
+                    parameterizedPropertyRenameDifferences.add(removalInfo.difference());
+                    parameterizedPropertyRenameDifferences.add(matchingAddition.difference());
+                }
+            }
+        }
+
+        if (serviceIdsWithMatchingAdditions.isEmpty() && parameterizedPropertyRenameDifferences.isEmpty()) {
+            return EnvironmentalChangeContext.empty();
+        }
+
+        return new EnvironmentalChangeContext(serviceIdsWithMatchingAdditions, parameterizedPropertyRenameDifferences);
+    }
+
+    public static boolean isControllerServiceCreatedForNewProperty(final FlowDifference difference, final EnvironmentalChangeContext context) {
```

Comment:
Can `context` be null? Similar to other public methods, I recommend using `Objects.requireNonNull()`